fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/group/ChunkReader.java

```diff
@@ -0,0 +1,328 @@
+                pageNum = 1;
+                chunkData.setPartNum(1);
+                indexOfCurrentResourceType++;
+            }
+        }
+        if (fhirTenant == null) {
+            fhirTenant = Constants.DEFAULT_FHIR_TENANT;
+            logger.info("readItem: Set tenant to default!");
+        }
+        if (fhirDatastoreId == null) {
+            fhirDatastoreId = Constants.DEFAULT_FHIR_TENANT;
+            logger.info("readItem: Set DatastoreId to default!");
+        }
+        if (fhirSearchPageSize != null) {
+            try {
+                pageSize = Integer.parseInt(fhirSearchPageSize);
+                logger.fine("readItem: Set page size to " + pageSize + ".");
+            } catch (Exception e) {
+                logger.warning("readItem: Set page size to default(" + Constants.DEFAULT_SEARCH_PAGE_SIZE + ").");
+            }
+        }
+
+        if (patientMembers == null) {
+            Group group = findGroupByID(fhirTenant, fhirDatastoreId, fhirSearchPatientGroupId);
+            patientMembers = new ArrayList<>();
+            // List for the group and sub groups in the expansion paths, this is used to avoid dead loop caused by circle reference of the groups.
+            HashSet<String> groupsInPath = new HashSet<>();
+            expandGroup2Patients(fhirTenant, fhirDatastoreId, group, patientMembers, groupsInPath);
+        }
+        List<Member> patientPageMembers = patientMembers.subList((pageNum - 1) * pageSize,
+                pageNum * pageSize <= patientMembers.size() ? pageNum * pageSize : patientMembers.size());
+        pageNum++;
+
+        if (chunkData == null) {
+            chunkData = new TransientUserData(pageNum, null, new ArrayList<PartETag>(), 1);
+            chunkData.setIndexOfCurrentResourceType(0);
+            jobContext.setTransientUserData(chunkData);
+        } else {
+            chunkData.setIndexOfCurrentResourceType(indexOfCurrentResourceType);
+            chunkData.setPageNum(pageNum);
+        }
+        chunkData.setLastPageNum((patientMembers.size() + pageSize -1)/pageSize );
+
+        if (!patientPageMembers.isEmpty()) {
+            logger.fine("readItem: loaded patients number - " + patientMembers.size());
+            fillChunkDataBuffer(patientPageMembers);
+        } else {
+            logger.fine("readItem: End of reading!");
+        }
+
+        return patientPageMembers;
+    }
+
+    @Override
+    public void open(Serializable checkpoint) throws Exception {
+        if (checkpoint != null) {
+            CheckPointUserData checkPointData = (CheckPointUserData) checkpoint;
+            pageNum = checkPointData.getPageNum();
+            indexOfCurrentResourceType = checkPointData.getIndexOfCurrentResourceType();
+            jobContext.setTransientUserData(TransientUserData.fromCheckPointUserData(checkPointData));
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+
```

Comment:
maybe add a logger.fine for debug purposes?

Reply:
also chunkData.getBufferStream() <-- who is response to close this?

Reply:
haha, good question. Actually Closing a ByteArrayOutputStream has no effect per java document, actually the function is empty in codes. it's handled by garbage collector after no reference to it.  and the same ByteArrayOutputStream is used the whole life circle or the batch job.

Reply:
OK, I think adding a comment to that effect is good thing in the code.